Sources/SourceKitLSP/SourceKitServer.swift

```diff
@@ -548,65 +565,65 @@ extension SourceKitServer: MessageHandler {
 
       switch request {
       case let request as Request<InitializeRequest>:
-        await self.initialize(request)
+        await self.handleRequest(request, handler: self.initialize)
       case let request as Request<ShutdownRequest>:
-        await self.shutdown(request)
+        await self.handleRequest(request, handler: self.shutdown)
       case let request as Request<WorkspaceSymbolsRequest>:
-        await self.workspaceSymbols(request)
+        await self.handleRequest(request, handler: self.workspaceSymbols)
       case let request as Request<PollIndexRequest>:
-        await self.pollIndex(request)
+        await self.handleRequest(request, handler: self.pollIndex)
       case let request as Request<ExecuteCommandRequest>:
-        await self.executeCommand(request)
+        await self.handleRequest(request, handler: self.executeCommand)
       case let request as Request<CallHierarchyIncomingCallsRequest>:
-        await self.incomingCalls(request)
+        await self.handleRequest(request, handler: self.incomingCalls)
       case let request as Request<CallHierarchyOutgoingCallsRequest>:
-        await self.outgoingCalls(request)
+        await self.handleRequest(request, handler: self.outgoingCalls)
       case let request as Request<TypeHierarchySupertypesRequest>:
-        await self.supertypes(request)
+        await self.handleRequest(request, handler: self.supertypes)
       case let request as Request<TypeHierarchySubtypesRequest>:
-        await self.subtypes(request)
+        await self.handleRequest(request, handler: self.subtypes)
       case let request as Request<CompletionRequest>:
-        await self.withLanguageServiceAndWorkspace(for: request, requestHandler: self.completion, fallback: CompletionList(isIncomplete: false, items: []))
+        await self.handleRequest(for: request, requestHandler: self.completion, fallback: CompletionList(isIncomplete: false, items: []))
       case let request as Request<HoverRequest>:
-        await self.withLanguageServiceAndWorkspace(for: request, requestHandler: self.hover, fallback: nil)
+        await self.handleRequest(for: request, requestHandler: self.hover, fallback: nil)
```

Comment:
Presumably the fact that we're now awaiting for results means that e.g if we've sent off a completion request we'll now need to wait for a response before we can process any notifications that require a barrier. Is that expected? Is this something that will be helped by cancellation?

Reply:
We discussed this offline. We should have a more granular tracking of dependencies in `AsyncQueue`. I’ve got some ideas of how to do this and will create a follow-up PR in the next couple of weeks.